### ISSUE TITLE
Support view *.cmi, *.cmt, *.cmo, *.bc and *.pp.ml files

### ISSUE THREAD

**Issue by hackwaly** [ocamllabs/vscode-ocaml-platform#505]
VS Code supports define custom editor. It would be useful that we can use it to show dump info of these files.

Files `*.cmi`, `*.cmt`, `*.cmo`, `*.bc`... can dump by `ocamlobjdump`.
And files `*.pp.ml` produced by dune can also dump by `ocamlc -dsource`.

Here's demo custom editor:
![image](https://user-images.githubusercontent.com/540298/105319418-ced84480-5bff-11eb-9170-0f598d38b004.png)

**Comment by rgrinberg** [comment]
> Files *.cmi, *.cmt, *.cmo, *.bc... can dump by ocamlobjdump.

Sounds like a good idea to me.

> And files *.pp.ml produced by dune can also dump by ocamlc -dsource.

Does this require `-I` and other flags as well? If so, this is more complex and will require integration from dune.

In any case, this is a useful feature, so if someone wants to work on it, a pr would be welcome.

**Comment by tatchi** [comment]
Hey @hackwaly,

> Files *.cmi, *.cmt, *.cmo, *.bc... can dump by ocamlobjdump

Did you mean `ocamlobjinfo` by any chance?

**Comment by tatchi** [comment]
I created a little POC. Is that what you had in mind?

https://user-images.githubusercontent.com/5595092/143494208-285fda4d-f44d-42bf-b2b2-8ca30c7f4755.mov

**Comment by rgrinberg** [comment]
@tatchi looks solid to me.

**Comment by mnxn** [comment]
In the ocaml source code there is both [objinfo.ml](https://github.com/ocaml/ocaml/blob/4.13/tools/objinfo.ml) and [dumpobj.ml](https://github.com/ocaml/ocaml/blob/4.13/tools/dumpobj.ml) in the tools directory. 

`objinfo.ml` seem to be built into `ocamlobjinfo` and included by default, but `dumpobj` is not.

It might be useful to use `dumpobj` because it shows the bytecode instructions (according to it's source code). `ocamlobjinfo` seems pretty limited in the information it shows.

**Comment by mnxn** [comment]
I had to build `dumpobj` from source with `make dumpobj` in `tools`. 
Here's a sample of what it can output:

<details>
  <summary> Excerpt </summary>
  
```
$ ./dumpobj.opt ./dumpobj.cmo
## start of ocaml dump of "./dumpobj.cmo"
       0  BRANCH 2184
File "dumpobj.ml", line 108, characters 26-52:
       2  ACC0
       3  PUSHENVACC2
       4  APPLY1
File "dumpobj.ml", line 108, characters 26-39:
       5  GETGLOBAL <0>(<11>("; ", 0), "; ")
       7  PUSHGETGLOBALFIELD Stdlib__printf, 1
      10  APPTERM1 2
File "dumpobj.ml", line 103, characters 30-63:
      12  GETGLOBAL <0>(<11>(", ", 0), ", ")
      14  PUSHGETGLOBALFIELD Stdlib__printf, 1
      17  APPLY1
File "dumpobj.ml", line 103, characters 30-41:
      18  ACC0
      19  PUSHENVACC2
      20  APPTERM1 2
File "dumpobj.ml", line 86, characters 29-877:
      22  ACC0
      23  SWITCH 
        tag 0 -> 29
        tag 1 -> 107
        tag 2 -> 175
        tag 3 -> 199
      29  ACC0
      30  GETFIELD0
      31  PUSHACC0
      32  SWITCH 
        tag 0 -> 41
        tag 1 -> 51
        tag 2 -> 61
        tag 3 -> 71
        tag 4 -> 77
        tag 5 -> 87
        tag 6 -> 97
      41  ACC0
      42  GETFIELD0
      43  PUSHACC0
File "dumpobj.ml", line 87, characters 31-44:
      44  PUSHGETGLOBAL <0>(<4>(0, 0, 0, 0), "%d")
      46  PUSHGETGLOBALFIELD Stdlib__printf, 1
      49  APPTERM2 5
      51  ACC0
      52  GETFIELD0
      53  PUSHACC0
File "dumpobj.ml", line 91, characters 32-45:
      54  PUSHGETGLOBAL <0>(<1>(0), "%C")
      56  PUSHGETGLOBALFIELD Stdlib__printf, 1
      59  APPTERM2 5
      61  ACC0
      62  GETFIELD0
      63  PUSHACC0
File "dumpobj.ml", line 89, characters 42-55:
      64  PUSHGETGLOBAL <0>(<3>(0, 0), "%S")
      66  PUSHGETGLOBALFIELD Stdlib__printf, 1
      69  APPTERM2 5
      71  ACC0
      72  GETFIELD0
      73  PUSHACC0
File "dumpobj.ml", line 88, characters 33-46:
      74  PUSHENVACC2
      75  APPTERM1 4
      77  ACC0
      78  GETFIELD0
      79  PUSHACC0
File "dumpobj.ml", line 92, characters 33-48:
      80  PUSHGETGLOBAL <0>(<5>(0, 0, 0, <12>('l', 0)), "%ldl")
      82  PUSHGETGLOBALFIELD Stdlib__printf, 1
      85  APPTERM2 5
      87  ACC0
      88  GETFIELD0
      89  PUSHACC0
```

</details>

I think the `ocamlobjinfo` output can work as the default view for these files since it supports all 4 files (`.cmi`, `.cmt`, `.cmo`, and `.bc`). `dumpobj` is useful for dissembling `.cmo` and `.bc` files but it doesn't do anything for `.cmi` and `.cmt`.

In the long term I would like to add support for `dumpobj` as well, but it seems much harder since it is not installed into a switch by default. As far as I know, there are no opam packages that provide `dumpobj` either. I will try to investigate this later.

@tatchi That POC looks good for now. Feel free to send a PR when it's ready.

**Comment by rgrinberg** [comment]
@mnxn when is inspecting the bytecode ever useful? It's both unreadable and pointless for optimization as bytecode is slow anyway.

**Comment by mnxn** [comment]
> when is inspecting the bytecode ever useful?

Not sure. I've personally never needed to inspect it.

> It's both unreadable and pointless for optimization as bytecode is slow anyway.

Fair enough. I made an assumption based on my JVM bytecode experience. I guess it's really a different situation.

The `ocamlobjinfo` output is definitely more useful in most cases.

**Comment by tatchi** [comment]
Can this one be closed now that #798 has been merged ?

**Comment by ulugbekna** [comment]
Thanks @tatchi :-)
